kore/src/Kore/Step/SMT/Sorts.hs

```diff
@@ -0,0 +1,416 @@
+
+    constructorArgToSMT
+        :: Text -> Int -> Sort Object -> Maybe (Text, SMT.SExpr)
+    constructorArgToSMT constrName index constrSort = do
+        translatedSort <- translateSort constrSort
+        return (constrName <> pack (show index), translatedSort)
+
+parseNoJunkAxioms
+    :: IndexedModule
+        param
+        (KorePattern
+            Domain.Builtin
+            Variable
+            (Unified (Valid (Unified Variable)))
+        )
+        Symbol
+        Attribute.Axiom
+    -> [SMTDataType]
+parseNoJunkAxioms indexedModule =
+    filter (\ SMTDataType{constructors} -> not (null constructors))
+    $ mapMaybe
+        parseSMTDataType
+        (recursiveIndexedModuleAxioms indexedModule)
+
+parseSMTDataType
+    ::  ( Attribute.Axiom
+        , SentenceAxiom param VerifiedKorePattern
+        )
+    -> Maybe SMTDataType
+parseSMTDataType (attributes, SentenceAxiom {sentenceAxiomPattern})
+  | Axiom.Constructor.isConstructor (Attribute.constructor attributes)
+  = case fromKorePattern Object sentenceAxiomPattern of
+    Left err -> (error . unlines)
+        [ "Unexpected error transforming kore pattern to pure pattern:"
+        , " all patterns should be pure."
+        , "err=" ++ show err
+        , "pattern=" ++ show sentenceAxiomPattern
+        ]
+    Right patt -> parseSMTDataTypePattern patt
+  | otherwise = Nothing
+
+parseSMTDataTypePattern :: CommonStepPattern Object -> Maybe SMTDataType
+parseSMTDataTypePattern (Bottom_ sort) =
+    return SMTDataType
+        { sort
+        , constructors = []
+        }
+parseSMTDataTypePattern (Or_ _ first second) = do
+    constructor <- parseSMTConstructor first
+    SMTDataType {sort, constructors} <- parseSMTDataTypePattern second
+    return SMTDataType
+        { sort
+        , constructors = constructor : constructors
+        }
+parseSMTDataTypePattern _ = Nothing
+
+parseSMTConstructor
+    :: CommonStepPattern Object -> Maybe SMTConstructor
+parseSMTConstructor patt =
+    case parsedPatt of
+        App_ symbol children -> do
+            childVariables <-
+                checkOnlyQuantifiedVariablesOnce quantifiedVariables children
+            buildConstructor symbol childVariables
+        _ -> Nothing
+  where
+    (quantifiedVariables, parsedPatt) = parseExists patt
+
+    parseExists
+        :: CommonStepPattern Object
+        -> (Set.Set (Variable Object), CommonStepPattern Object)
+    parseExists (Exists_ _ variable child) =
+        (Set.insert variable childVars, unquantifiedPatt)
+      where
+        (childVars, unquantifiedPatt) = parseExists child
+    parseExists unquantifiedPatt = (Set.empty, unquantifiedPatt)
+
+    checkOnlyQuantifiedVariablesOnce
+        :: Set.Set (Variable Object)
+        -> [CommonStepPattern Object]
+        -> Maybe [Variable Object]
+    checkOnlyQuantifiedVariablesOnce
+        allowedVars
+        []
+      | Set.null allowedVars = Just []
+      | otherwise = Nothing
+    checkOnlyQuantifiedVariablesOnce
```

Comment:
What is this checking?

Reply:
I'm doing some sanity checks that I'm parsing what I think I should be parsing.

In particular, this is part of a check that axiom is an `or` of patterns of the type `exists x1 x2 .. xn . sigma(y1, y2, ... ym)`, and this function checks that `m=n` and `{x1, x2, ..., xn} = {y1, y2, ..., yn}`.